# Crear blueprint

Per verificar que datasets.ods i el notebook estan a la llista

In [52]:
import os, sys
print("Python executable:", sys.executable)     # mostra el Python que corre (ha de ser el env activat)
print("Directori de treball:", os.getcwd())    # carpeta del repo
print("Llistat d'arxius a la carpeta:")
for f in sorted(os.listdir('.')):
    print(" -", f)

Python executable: c:\Users\UX433\miniconda3\envs\myenv\python.exe
Directori de treball: c:\Users\UX433\OneDrive\Escritorio\networkalignment
Llistat d'arxius a la carpeta:
 - .git
 - A2_blueprint.npy
 - A_test_equiv.npy
 - C-el_k4_Synthetic- No groups - Simoultenously.ipynb
 - Hs_base0.npy
 - Hs_base1.npy
 - Hs_synthetic_paper.npy
 - L_latent_from_pq.npy
 - L_latent_repo_exact_errors.npy
 - Matriz_epochs_A.pickle
 - Syntetic_succesive_tot_0.pickle
 - codienbrut.ipynb
 - crear_blueprint.ipynb
 - datasets.ods
 - energies_base0.npy
 - energies_base1.npy
 - energy_pipeline_summary.npy
 - environment.yml
 - synthetic_paper.npy
 - synthetic_paper_0.npy
 - synthetic_repo_errors.npy


In [53]:
import itertools
import pickle
import matplotlib.pyplot as plt
import numpy as np
import math
import random
import os
import time
from numba import jit, njit
from numba.types import bool_, int_, float32
from math import comb
from copy import deepcopy
from tqdm import tqdm
import networkx as nx
import pandas as pd
from collections import defaultdict
from itertools import permutations

### Llegir fulls de càlcul 

In [54]:
### Open the files
fulls = 8
#n_grupo = 4 - ara no estic fent servir grups

d = {} #crea un diccionari per emmagatzemar DataFrames(taules de memòria)
D = "Dataset" #prefix del nom de les fulles — al excel es diuen Dataset1, Dataset2, ...
for i in range(0,fulls): #recorro 0...7
    d["group" + str(i)] = pd.read_excel("datasets.ods", sheet_name=D+str(i+1)) #el que al diccionari serà gorup 0, sera el meu Dataset1
    # hem assignat al diccionari: d["group0"] = <taula Dataset1> 
    print('Read:',i, 'size', d["group" + str(i)].shape) #volem imprimir el nre de files i columnes

Read: 0 size (227, 183)
Read: 1 size (227, 183)
Read: 2 size (227, 183)
Read: 3 size (227, 183)
Read: 4 size (227, 183)
Read: 5 size (227, 183)
Read: 6 size (227, 183)
Read: 7 size (227, 183)


### Passar a matriu i binaritzem

Consideracions a tenir en compte. Primer, treiem les 3 primeres files i columnes de cada full de càlcul perquè a l'excel contenen etiquetes, no dades. En segon lloc, tenim 227 diles i 183 columnes. El que fem és forçar una matriu quadrada nre.filesxnre.files, de manera que, com la matriu inicial eren zeros, copio les columnes existets i deixo la resta plenes de zeros. Això podria donar error. NOTA: Els datasets compten la primera columna com títols.

In [55]:
## Hipermatrix M with the data (Only the synapses) 
rows = d["group" + str(i)].shape[0] - 3
columns = d["group" + str(i)].shape[1] - 3 #les tres primeres files i columnes no contenen dades

M = np.zeros((fulls,rows,columns))
for i in range(0, fulls):
    Data = d['group' + str(i)]
    M[i,:,:] = Data.iloc[3:,3:]
    
## Since we work with same number of nodes, we want them equal and square (zeros when no connections)
M_square = np.zeros((fulls, rows, rows))#creo matriu quadrada
M_square[:,:, 0:columns] = M[:,:,:] #copies les columnes existents (0..179) per deixar zeros en les restants (180..223).

## Binarization: No weights
#Lo queremos BINARIO, ignorando su peso (Luego pensar cómo se haría con el peso)
M_square_bin = np.zeros((fulls,rows,rows))
for i in range(fulls):
    for j in range(rows):
        for k in range(columns):
            if (M[i,j,k] >= 1):
                M_square_bin[i,j,k] = 1
                
Nx = rows # Number of nodes (we imposed rows == columns)


### Comprovació de l’estructura dels fulls de càlcul

Vull assegurar-me de que la neteja dels Datasets (primeres 3 files i columnes) està feta correctament. Nota: Els Datasets prenen la primera fila com un Títol  ('Pre', en aquest cas). Per tant, realment n'estic esborrant 4.

In [56]:
# inspecció ràpida d'un full de càlcul (group 7-->Dataset8)
df7 = d['group7']
print("Shape original group7:", df7.shape)
print("\nColumn names (first 12):\n", df7.columns[:12])
print("\nPrimeres 8 files i 8 columnes (sense transformar):")
print(df7.iloc[:8, :8])

Shape original group7: (227, 183)

Column names (first 12):
 Index(['Unnamed: 0', 'Unnamed: 1', 'Unnamed: 2', 'Pre', 'Unnamed: 4',
       'Unnamed: 5', 'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9',
       'Unnamed: 10', 'Unnamed: 11'],
      dtype='object')

Primeres 8 files i 8 columnes (sense transformar):
  Unnamed: 0 Unnamed: 1 Unnamed: 2      Pre Unnamed: 4 Unnamed: 5 Unnamed: 6  \
0        NaN        NaN        NaN  Sensory        NaN        NaN        NaN   
1        NaN        NaN        NaN     ADFL       ADFR       ADLL       ADLR   
2        NaN        NaN        NaN      NaN        NaN        NaN        NaN   
3       Post    Sensory       ADFL        0          0          0          0   
4        NaN        NaN       ADFR        0          0          0          0   
5        NaN        NaN       ADLL        0          0          0          0   
6        NaN        NaN       ADLR        0          0          0          0   
7        NaN        NaN       AFDL      

## Escollir quin és el dataset A2

Sabem (per la informació del paper) que A2 té aproximadament 2186 arestes (edges), comprovem si el Dataset8 concorda amb aquestes dades, mirem cada dataset (recordem que group0-->Dataset1, group1-->Dataset2, ...).

NOTA: A més sospitem que el Dataset8 és el que conté les dades de A2 perquè al repo de la Teresa es pren A2 com a Blueprint fent M_square_bin[-1] que correspon a l'últim element de l'array, és a dir, M_square_bin[7], que correspon a group7--> Dataset8

### Coprovació de que Dataset8 correspon a A2

In [57]:
# Veure quants edges té cada dataset
for i in range(8):
    print(i, M_square_bin[i].sum())

0 775.0
1 986.0
2 1012.0
3 1136.0
4 1515.0
5 1525.0
6 2202.0
7 2186.0


## Guardar A2 com blueprint

Volem crear i guardar el blueprint A2 (Dataset8 = group7 = M_square_bin[-1])


In [58]:
#Seleccionem A2, és a dir, el dataset8
A2=M_square_bin[-1].copy()
#Fem comprovacions
print("A2 creat")
print("Files i columnes A2", A2.shape)
print("Arestes", int(A2.sum()))
#guardem
np.save("A2_blueprint.npy", A2)
print("Guardat com A2_blueprint")

A2 creat
Files i columnes A2 (224, 224)
Arestes 2186
Guardat com A2_blueprint


# Generar xarxes amb error

Dues maneres, fixar un nombre d'errors (com es fa al codi de la Teresa) o fer servir p i q (probabilitats de copiar malament un link/no-link). Farem totes dues

### ESBORRAR AIXÒ: Veure d'on sortien els errors q feia servir Teresa

In [59]:
L = np.load("A2_blueprint.npy")
Nx = L.shape[0]
edges = int(L.sum())
zeros = Nx*Nx - edges
e_error = 327
z_error = 327
q_equiv = e_error / edges
p_equiv = z_error / zeros
print("edges", edges, "zeros", zeros)
print("q_equiv =", q_equiv)
print("p_equiv =", p_equiv)

edges 2186 zeros 47990
q_equiv = 0.1495882891125343
p_equiv = 0.00681391956657637


Hem vist que p i q del repositori son practicament els mateixos del paper

### Càlcul de nombre d'errors equivalents per p i q

In [60]:
# Carrega el blueprint A2 desat
L = np.load("A2_blueprint.npy")

Nx = L.shape[0]

print("Blueprint A2 carregat correctament.")
print("Mida:", L.shape)
print("Nombre d'enllaços (1’s):", int(L.sum()))


Blueprint A2 carregat correctament.
Mida: (224, 224)
Nombre d'enllaços (1’s): 2186


In [61]:
# Errors deduits dels valors del paper
edges_A2 = L.sum()
zeros_A2 = Nx*Nx - edges_A2

q_paper = 0.15
p_paper = 0.007

errors_esborrar = int(q_paper * edges_A2)
errors_crear = int(p_paper * zeros_A2)

print(errors_esborrar, errors_crear)


327 335


No són exactament els mateixos, suposo que això es deu a què al repo es fan servir p i q que no quadren exactament amb els del paper. 

(Esborrar això) ara faré el mateix però amb els p i q que he deduit del codi de la teresa, per veure si em surt el mateix nre, d'errors que ella ha fet servir (327,327) i comprovar validesa de l'anterior

In [62]:
# Errors deduits dels valors del paper
edges_A2 = L.sum()
zeros_A2 = Nx*Nx - edges_A2

errors_esborrar_teresa = int(q_equiv * edges_A2)
errors_crear_teresa = int(p_equiv * zeros_A2)

print(errors_esborrar_teresa, errors_crear_teresa)


327 327


Sí, el nombre d'errors quadra

# Generar xarxes amb nombre d'errors

Ara estem generant xarxes com es feia al repositori, amb el mateix nombre d'errors exacte

In [63]:
'''# === Generar xarxes amb errors exactes (estil del repo) ===

K = 2             # nombre de xarxes a generar
e_error = 327       # errors esborrant enllaços
z_error = 327       # errors creant enllaços

# Carregar blueprint si cal
if 'L' not in globals():
    print("No hi ha L a la memòria, carregant A2_blueprint.npy ...")
    L = np.load("A2_blueprint.npy")

Nx = L.shape[0]
print("Blueprint A2 carregat: shape", L.shape, "edges:", int(L.sum()))

# Preparar output
A_repo = np.zeros((K, Nx, Nx), dtype=int)
rng = np.random.default_rng(42)  # llavor fixa

edges_total = int(L.sum())
zeros_total = Nx * Nx - edges_total

print("Edges A2:", edges_total, "Zeros A2:", zeros_total)

# Comprovacions
if e_error > edges_total:
    raise ValueError(f"e_error ({e_error}) > nombre d'enllaços ({edges_total})")
if z_error > zeros_total:
    raise ValueError(f"z_error ({z_error}) > nombre de zeros ({zeros_total})")

for m in range(K):
    A = L.copy().astype(int)
    A_flat = A.reshape(Nx*Nx)

    ones_idx = np.where(A_flat == 1)[0]
    zeros_idx = np.where(A_flat == 0)[0]

    flip_ones = rng.choice(ones_idx, size=e_error, replace=False)
    flip_zeros = rng.choice(zeros_idx, size=z_error, replace=False)

    A_flat[flip_ones] = 0
    A_flat[flip_zeros] = 1

    A = A_flat.reshape(Nx, Nx)
    A_repo[m] = A

    errors_removed = edges_total - np.sum((A == 1) & (L == 1))
    errors_added = np.sum((A == 1) & (L == 0))

    print(f"Xarxa {m}: Edges = {int(A.sum())}, Errors esborrant = {errors_removed}, Errors creant = {errors_added}")

# Guardar a disc
out_name = "synthetic_repo_errors.npy"
np.save(out_name, A_repo)
print("\nGuardades les xarxes amb errors exactes a:", out_name)
'''

'# === Generar xarxes amb errors exactes (estil del repo) ===\n\nK = 2             # nombre de xarxes a generar\ne_error = 327       # errors esborrant enllaços\nz_error = 327       # errors creant enllaços\n\n# Carregar blueprint si cal\nif \'L\' not in globals():\n    print("No hi ha L a la memòria, carregant A2_blueprint.npy ...")\n    L = np.load("A2_blueprint.npy")\n\nNx = L.shape[0]\nprint("Blueprint A2 carregat: shape", L.shape, "edges:", int(L.sum()))\n\n# Preparar output\nA_repo = np.zeros((K, Nx, Nx), dtype=int)\nrng = np.random.default_rng(42)  # llavor fixa\n\nedges_total = int(L.sum())\nzeros_total = Nx * Nx - edges_total\n\nprint("Edges A2:", edges_total, "Zeros A2:", zeros_total)\n\n# Comprovacions\nif e_error > edges_total:\n    raise ValueError(f"e_error ({e_error}) > nombre d\'enllaços ({edges_total})")\nif z_error > zeros_total:\n    raise ValueError(f"z_error ({z_error}) > nombre de zeros ({zeros_total})")\n\nfor m in range(K):\n    A = L.copy().astype(int)\n    A_f

# Generar xarxes amb p i q

Ara volem comprovar el nombre d'enllaços que s'han creat (p·zeros ≈ 0.007·48000 ≈ 330–350 nous) i esborrat (q·edges ≈ 0.15·2186 ≈ 327–330 enllaços)

El que estic fent aquí no és modificar el blueprint amb errors, no reconstruir des de 0. No estic recreant la matriu probabilisticament partint d'una matriu de 0s. 
Aqui la matriu L comença tocada i només altero alguns elements.

In [64]:
num_networks = 2
A_test = np.zeros((num_networks, Nx, Nx)) #crea un array 3D ple de zeros

p = p_paper
q = q_paper

np.random.seed(1) #fixo la llavor de les xarxes random generades pq sino cada execucio em genera xarxes diferents

for k in range(num_networks):
    A_copy = L.copy() #deep copy de L, però no comparteixen memòria. modificar a.copy no modifica L
    for i in range(Nx):
        for j in range(Nx):
            if L[i,j] == 1:
                # enllaç existent -> pot ser esborrat
                if np.random.rand() < q: #genero aleatori entre 0 i 1
                    A_copy[i,j] = 0
            else:
                # posició buida -> pot crear-se un fals
                if np.random.rand() < p:
                    A_copy[i,j] = 1
    A_test[k,:,:] = A_copy #guardo cada xarxa k generada.
    # per cada k (ie per cada xarxa), es comença de zero fent una còpia nova i completa de L.


print("Xarxes generades amb p i q del paper:")
print(A_test.shape)
np.save("synthetic_paper.npy", A_test)


Xarxes generades amb p i q del paper:
(2, 224, 224)


Si ho vull fer tal com em va dir la marta, emplenant una matriu feta de 0s seria aixi.

In [65]:

A_test_0 = np.zeros((num_networks, Nx, Nx))
np.random.seed(1)

for k in range(num_networks):
    A_copy_0 = np.zeros_like(L)   # matriu tota de 0

    for i in range(Nx):
        for j in range(Nx):

            if L[i, j] == 1:
                # Enllaç existent
                if np.random.rand() < q:
                    A_copy_0[i, j] = 0        # No copio (error d'esborrat)
                else:
                    A_copy_0[i, j] = 1        # Copio
            else:
                # No enllaç
                if np.random.rand() < p:
                    A_copy_0[i, j] = 1        # Fals positiu
                else:
                    A_copy_0[i, j] = 0        # Copio (0)

    A_test_0[k] = A_copy_0
print("Xarxes generades amb p i q del paper partint d'una de 0s (recrear probabilisticament):")
print(A_test_0.shape)
np.save("synthetic_paper_0.npy", A_test_0)

Xarxes generades amb p i q del paper partint d'una de 0s (recrear probabilisticament):
(2, 224, 224)


In [66]:
# Comprovació del nombre d’enllaços final
original_edges = int(L.sum())
zeros_original = Nx*Nx - original_edges

print("Blueprint (A2) edges:", original_edges)
print("Blueprint zeros:", zeros_original)

for k in range(num_networks):
    edges_new = int(A_test[k].sum())
    print(f"\nXarxa {k}:")
    print("  Edges:", edges_new)

    # errors equivalents
    errors_esborrats = original_edges - np.sum((A_test[k] == 1) & (L == 1))
    errors_creats = np.sum((A_test[k] == 1) & (L == 0))

    print("  Errors esborrant:", errors_esborrats)
    print("  Errors creant:", errors_creats)


Blueprint (A2) edges: 2186
Blueprint zeros: 47990

Xarxa 0:
  Edges: 2207
  Errors esborrant: 320
  Errors creant: 341

Xarxa 1:
  Edges: 2174
  Errors esborrant: 333
  Errors creant: 321


## Funcions d'energia

Aquí enganxo les funcions d'energia del repositori del paper, del notebook "C-el_k4_Synthetic-No groups...", ja que son les mateixes que he de fer servir.

No considerem grups, per tant:

In [67]:
n_groups = 1 #considerem que els nodes pertanyen a un únic grup
start_groups = np.zeros(1) #indica on comença cada grup
end_groups = np.zeros(1) + Nx #indica on acaba
size_groups = np.zeros(1) + Nx

start_groups, end_groups, size_groups = start_groups.astype(int), end_groups.astype(int), size_groups.astype(int)

Aquestes son les funcions d'energia tal qual com estan al repo del paper

In [68]:
@jit(nopython = True)
def hamiltonian_prob(Edges_NoL, Edges_L, overlap_0, overlap_1, alpha, beta):

    A_1 = overlap_1 + alpha
    B_1 = (Edges_L - overlap_1 + beta)
    C_1 = Edges_L + alpha + beta
    
    A_0 = overlap_0 + alpha
    B_0 = (Edges_NoL - overlap_0 + beta)
    C_0 = Edges_NoL + alpha + beta
    
    #  [ math.lgamma(n+1) == log(n!) ]
    H1 = math.lgamma(A_1)+ math.lgamma(B_1) - math.lgamma(C_1) 
    H0 = math.lgamma(A_0)+ math.lgamma(B_0) - math.lgamma(C_0) 
    
    H = -(H1 + H0)
    return H
    
    
@jit(nopython=True)
def overlap_total_prob(L_f, A_f, P_inv_f):
    Nx = L_f.shape[0]
    Ny = L_f.shape[1]
    K = A_f.shape[0]
    
    ovlp_0 = np.zeros((K))
    ovlp_1 = np.zeros((K))
    for k in range(0,K):
        for f in range(0,Ny): 
            for c in range(0,Nx):
                p_f=int(P_inv_f[k,f])
                p_c=int(P_inv_f[k,c])  
                
                valor_L, valor_A = L_f[f,c], A_f[k,p_f,p_c]
                
                ovlp_0[k] = ovlp_0[k] + (1-valor_L)*(1-valor_A )
                ovlp_1[k] = ovlp_1[k] + valor_L*valor_A
                
                
    ovlp_1 = int(sum(ovlp_1))
    ovlp_0 = int(sum(ovlp_0))
    return ovlp_0, ovlp_1


@jit(nopython=True) # The blueprint is the average of the observations (taking into account the mapping)  
def L_wiring(A_f, P_inv_f):
    
    Nx = A_f.shape[1]
    Ny = A_f.shape[2]
    K = A_f.shape[0]
    L_new_f = np.zeros((Nx,Ny))
    
    for i in range(0,Nx):
        for j in range(0,Ny):
            for k in range(0,K):
        
                p1 = int(P_inv_f[k,i]) # Mapping of the observations
                p2 = int(P_inv_f[k,j]) # Mapping of the observations
                L_new_f[i,j] += A_f[k,p1,p2]
            valor_lnew=1/K* L_new_f[i,j]
            L_new_f[i,j] = round( valor_lnew ) 
            # If valor_lnew = 0, L=0 (we could establish L=1, but it is more probable to not have a connection)
    
    return L_new_f


#### Some algorithm for sorting
@jit(nopython=True)
def partition(array,  etiquetas, begin, end):
    pivot = begin
    for i in range(begin+1, end+1):
        if array[i] < array[begin]:
            pivot += 1
            array[i], array[pivot] = array[pivot], array[i]
            etiquetas[i], etiquetas[pivot] = etiquetas[pivot], etiquetas[i]
    array[pivot], array[begin] = array[begin], array[pivot]
    etiquetas[pivot], etiquetas[begin] = etiquetas[begin], etiquetas[pivot] 

    return pivot
@jit(nopython=True)
def quicksort(array, etiquetas, begin=0, end=None):
    if end is None:
        end = len(array) - 1
    if begin >= end: #To end
        return
    pivot = partition(array,  etiquetas, begin, end)
    
    #Order right and left
    quicksort(array, etiquetas, begin, pivot-1)
    quicksort(array,  etiquetas, pivot+1, end)
    
@njit 
# We inizialise the algorithm sorting by node degree, but also taking into account the group labels
def permu_groups(L_f, A_f, start_f): 
    Nx = L_f.shape[0]
    Ny = L_f.shape[1]
    K = A_f.shape[0]
    N_groups = len(start_f)
    
    
    # Mappings
    P_f = np.zeros((K,Nx)) #Mapping from L to A
    P_inv_f = np.zeros((K,Nx)) #Mapping from A to L
    P_new = np.zeros((K,Nx))
    
    # First we order the blueprint and after the observations
    # Blueprint:
    orden_L=np.zeros((Nx))
    for i in range(Nx):
        b=np.nonzero(L_f[i,:])
        orden_L[i] = b[0].size
        
    array_L = np.arange(Nx) #For the labels 
    quicksort(orden_L, array_L)
    
    # Now according to the group label
    array_L_labels = np.zeros((Nx))
    group_number = np.zeros((N_groups))
    for i in range(Nx):
        n_type = 0
        #Group
        while ( (array_L[i] >= start_f[n_type]) and (n_type < N_groups)):
            n_type += 1
       
        grupo = int(n_type-1)
        g = int(start_f[grupo])
        array_L_labels[g + int(group_number[grupo])] = array_L[i]
        group_number[grupo] += 1

    #Observations: 
    for i in range(0,K):
        orden_A = np.zeros((Nx))
        for i_orden in range(Nx):
            b = np.nonzero(A_f[i,i_orden,:])
            orden_A[i_orden] = b[0].size

        array_A = np.arange(Nx) #For the labels 
        quicksort(orden_A, array_A)
            
        # Now according to the group label
        array_A_labels = np.zeros((Nx))
        group_number = np.zeros((N_groups))
        
        for i_g in range(Nx):
            n_type = 0
            while ( (array_A[i_g] >= start_f[n_type]) and (n_type < N_groups) ):
                n_type += 1

            grupo = int(n_type-1)
            g = int(start_f[grupo])
            array_A_labels[ g + int(group_number[grupo])] = array_A[i_g]
            group_number[grupo] += 1
    

    # Now ordered together both label array  array_L = [0,1,2,3...]
        array_A_labels = array_A_labels[np.argsort(array_L_labels)]
        P_f[i,:] = array_A_labels
        
        for i_inv in range(0,Nx):
            for j_inv in range(0,Ny):
                if (P_f[i,i_inv] == j_inv):
                    P_inv_f[i, j_inv]=i_inv
    
    P_todo = np.zeros((2,K,Nx))
    P_todo[0,:,:] = P_f.copy() 
    P_todo[1,:,:] = P_inv_f.copy()

    return P_todo

 

## Generar 10 mostres

El que vull fer ara és agafar la blueprint i les dues xarxes generades i aplicar la identitat i les funcions d'energia. Després, fer un shuffle i aplicar energies. Generar 10 i fer historigrama

### pas 1

De moment agafo la latent, i les dues xarxes generades amb cert error i aplico identitat i funcions d'energia

In [73]:

#carrego fitxers
L = np.load("A2_blueprint.npy")                   # blueprint guardada
A_test = np.load("synthetic_paper.npy")           # 2 xarxes generades


# Construir Ao_stack i identitat (P_inv_id)
Ao_enters = A_test.astype(int)       # aqui convertim a int per si de cas havia quedat amb floats després de fer rand
P_inv_id = np.tile(np.arange(Nx), (K,1)).astype(int)   #fem k vectors identitat, perquè les funcions defineixen les permutacions com vectors de mmida NX, no matrius

print("Primeres 6 posicions de la primera fila del vector identitat, han de sortir en ordre", P_inv_id[0,:6])

#ara tinc:
#L → la blueprint (matriu binaria 224x224)
#A_test → les dues xarxes q he generat abans amb p i q(forma 2×224×224)
#P_inv_id → perm identitat (0,1,2,…,223), feta dos cops (per les dues xarxes)'''

#Cridar L_wiring sobre Ao_stack amb identitat, aixo em dona tot 0s, aixiq  ha d'estar malament
print("Aplico L_wiring")
L_nova = L_wiring(Ao_enters, P_inv_id)
print("L_nova shape:", L_nova.shape)

#comprovar si L_nova és igual a L
print("Similitud entre L y L_nova:", np.sum(L == L_nova), "de", Nx*Nx)
#si no surt 50176 de 50176, vol dir que L_wiring no està funcionant bé.


#Cridar overlap_total_prob amb la blueprint original (L) i Ao_stack amb identitat
print("Aplico overlap_total_prob")
ovlp0_identity, ovlp1_identity = overlap_total_prob(L, Ao_enters, P_inv_id)
print("ovlp0_identity =", ovlp0_identity)
print("ovlp1_identity =", ovlp1_identity) 

#un valor raonable seria tenir ovlp1_identity proper al nombre d'enllaços menys els errors esborrant, i ovlp0_identity proper al nombre de zeros menys els errors creant.
#com la majoria eres zeros, ovlp0_identity ha de ser molt gran, i ovlp1_identity ha de ser menor però proper al nombre d'enllaços menys els errors esborrant(menor que 2186).
#dona valors inconsistents, per tant alguna cosa no està funcionant bé.

#aplico hamiltonian_prob
# Calcular nombre d'enllaços i zeros a la blueprint
Nx = L.shape[0]
Edges_L = int(L.sum())                 # nombre de 1s de la blueprint, hauria de ser 2186
Edges_NoL = Nx*Nx - Edges_L            # nombre de 0s

print("Edges_L  =", Edges_L)
print("Edges_NoL =", Edges_NoL)

#Parametres alpha i beta
alpha = 1.0
beta  = 1.0
print("alpha =", alpha, " beta =", beta)

# Aplicar hamiltonian_prob
print("Aplico hamiltonian_prob")
H_identity = hamiltonian_prob(
    Edges_NoL,
    Edges_L,
    ovlp0_identity,
    ovlp1_identity,
    alpha,
    beta
)

print("H_identity =", H_identity)

#Guardar resultats, descomento quan estig segura que funciona
#np.save("H_identity.npy", np.array(H_identity))
#np.save("ovlp0_identity.npy", np.array(ovlp0_identity))
#np.save("ovlp1_identity.npy", np.array(ovlp1_identity))

print("PAS 1 complet correctament.")


Primeres 6 posicions de la primera fila del vector identitat, han de sortir en ordre [0 1 2 3 4 5]
Aplico L_wiring
L_nova shape: (224, 224)
Similitud entre L y L_nova: 49559 de 50176
Aplico overlap_total_prob
ovlp0_identity = 95318
ovlp1_identity = 3719
Edges_L  = 2186
Edges_NoL = 47990
alpha = 1.0  beta = 1.0
Aplico hamiltonian_prob
H_identity = -inf
PAS 1 complet correctament.
